[MIG] sale_order_project: Migration to 11.0

aitorbouzas opened the pull request:
Refactor of sale_order_project. It's almost the same behaviour but without the contract things.

oihane committed:
[IMP] <sale_project_base> renamed to <sale_order_project>
* related_project_id is computed
* permissions required for project creation

[IMP] <sale_order_project> Some suggested changes
* Version 1.0 is shown in Odoo as version 8.0.1.0
* Code cleaned

[IMP] <sale_order_project> Added tests

[IMP] <sale_order_project> Replaced camel case

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

sbidoul committed:
[FIX] remove en.po that was erroneously created by transbot

pedrobaeza committed:
[MIG] Rename manifest files

sbidoul committed:
remove obsolete .pot files [ci skip]

aitorbouzas committed:
[MIG] sale_order_project: Migration to 11.0

aitorbouzas committed:
[MIG] sale_order_project: Fixing space at the beginning of name

oihane commented:
@aitorbouzas you are not taking into account that there is a field `analytic_account_id` created by `sale` module, that's why the link to project is a computed field (in v8)

aitorbouzas commented:
@oihane Ok, the thing is the module in V8 is using a field that doesn't exist in V11 for ```sale.order``` model which is ```project_id```. I guess it is ```account_analytic_id``` now, right?

oihane commented:
Yes, that's it...

aitorbouzas committed:
[MIG] sale_order_project: Relation with project as computed

aitorbouzas commented:
@oihane is it ok now? Thanks for reviewing the module.
I don't know why runbot is saying that the manifest is missing OCA in author... It is there.

aitorbouzas committed:
[MIG] sale_order_project: Runbot fix missing author

Abranes commented:
Hello, notice that I have already done the migration to 10.0 (#377) where are solved the major migration problems like change openerp imports and xml tags.

I think it could be better if you start the migration from 10.0 PR (#377).

PD: Can someone give some love to my PR #377?

aitorbouzas commented:
@Abranes your module seems to just create a button that allows you create project from sale order... And the view is referencing an already created field in sale.order called ```related_project_id```. Where's that field added in V10?

Abranes commented:
@aitorbouzas that field is created in [sale module](https://github.com/odoo/odoo/blob/16fb8f0a7500218b7d53e89dcaaec3c5fb7d7eb5/addons/sale/models/sale.py#L134)

aitorbouzas commented:
@Abranes the problem is there's no such thing in V11

Abranes commented:
@aitorbouzas It's OK if you've to create this field. What I'm saying it's that you should start your PR from 10.0 branch not from 8.0 because there're things already done like replace openerp imports, remove data tags, remove api.one, etc...

anajuaristi commented:
I'm sorry but I think this module should not be migrated not from V10, nor from V8 but it should be completely redesigned and simplified due to Odoo core structure change related to projects/analytic account
Please let me a couple of days to think about something consistent, sinplified and redesign the aproach.
@oihane, we should talk about this tomorrow

SalahAdDin committed:
Merge pull request #1 from OCA/11.0

Update from original

SalahAdDin committed:
WIP [11.0][ADD][sale_order_project] Initial migration attempt

SalahAdDin committed:
:fire: applied 2to3 command.

SalahAdDin committed:
:sunny: improved spanish translation and added turkish translation.

SalahAdDin committed:
:fire: remove blank lines

SalahAdDin commented:
I was thinking it needs to be like `crm_project`:
![crmprojectboton](https://user-images.githubusercontent.com/5034215/43634503-f5e15ff6-9714-11e8-9389-405c52aaf9ad.png)
Just a button to create a new project.
![crmprojectpopupconmenu](https://user-images.githubusercontent.com/5034215/43634511-fced2c3a-9714-11e8-8276-c3d195938f6e.png)
A popup dialog with the option of creating a new project or choose once existing.

Simple but effective.

In our case, we must have a one-to-one relationship between Sales Order and Project.

@anajuaristi What do you think about it?
In addition, i saw that `mrp_project` module depends on `mrp_analytic`, i think it maybe has to see with your comment, i don't know.

@Abranes I did an attempt to in this [PR](https://github.com/OCA/project/pull/414).

SalahAdDin committed:
Merge branch '11.0-mig-sale_order_project' into 11.0

SalahAdDin committed:
:fire: improve manifest file.

SalahAdDin committed:
:fire: Refactored `Create Project` button.

SalahAdDin committed:
:fire: Removed deprecated license lines.

SalahAdDin committed:
:fire: Removed deprecated license lines.

SalahAdDin committed:
:sunny: Fix button ihneritance.

SalahAdDin committed:
:fire: Removed deprecated license lines.

SalahAdDin committed:
:fire: Reverted to the local approach.

SalahAdDin committed:
:fire: Fixed button icon.

SalahAdDin committed:
:fire: Reverted button type to  Object.

I'm sorry, it was a noob mistake.

SalahAdDin committed:
:fixed: sales.order has not project_id attribute.

SalahAdDin committed:
:fire: Fix deprecation warning: api.one

SalahAdDin committed:
:tada: Initial wizard attempt.

SalahAdDin committed:
:fire: reverted method to model file.

SalahAdDin committed:
:sunny: added wizard to init file.

SalahAdDin committed:
:fire: Commented inner code.

For now i don't know how can i handle it.

SalahAdDin committed:
:fire: Properly named modules.

SalahAdDin committed:
:fire: Properly named modules.

SalahAdDin committed:
:fire: correct action calling.

Until here all works fine, we can install it without problem.

We just need to handle the action.

SalahAdDin committed:
:fire: fixed button action.

SalahAdDin committed:
:fire: refoactored default get method.

SalahAdDin committed:
:sunny: added a success message and trying to handle the action.

SalahAdDin committed:
:fire: tweaks in wizard.

-  Reverted default get change.
- Added help text to the field.
- Added conditionals.

_Note:_ Right know only the first condition works: it does not matter if you choose a project, it ever calls the create project method.

SalahAdDin committed:
:fire: Refactored field format in python 3 way.

SalahAdDin committed:
:fire: Removed useless pop up.

SalahAdDin committed:
:fire: Format tweaks.

SalahAdDin committed:
:fire: refactored conditionals, 

Right now the method can link sale order and project succesfully.
But for empty field it ever takes the `else` stament. Why?

SalahAdDin committed:
:fire: fixed conditionals

Four hours in a simple `elif`.

SalahAdDin committed:
:fix: Refactored warnings' format.

SalahAdDin committed:
:fire: Update readme with usage section.

SalahAdDin committed:
:fire: Fixed company email.

SalahAdDin committed:
:tada: [WIP] Added button in Project to see all related Sale Orders

Create a new field for count.

SalahAdDin committed:
:sunny: Added button for project view.

SalahAdDin committed:
:sunny: Added button for project view. reference in manifest.

SalahAdDin committed:
:fire: beautify sale view template.

SalahAdDin committed:
:sunny: Added Sales Management as dependencie.

It has no sense to use buttons and others without the proper view.

SalahAdDin committed:
:sunny: Added properly action to the button.

SalahAdDin committed:
:sunny: [WIP] Creating the action for button.

SalahAdDin committed:
:fire: Added function to sale order count field.

-> Related projec id is now stored in data base.
-> Button now show the related sale orders number.

SalahAdDin committed:
:fire: Fixed sale order tree view problem.

SalahAdDin commented:
@anajuaristi One month. :D

SalahAdDin committed:
Merge pull request #1 from SalahAdDin/11.0

WIP [11.0][ADD][sale_order_project] Initial migration attempt

SalahAdDin commented:
Ok guys, this is our work in order to update this module.

But we have two or three problems:
First one, the `Project Overview` button is showing in wrong places and sometimes in the same place with  `Create Project` button:
![aquiaparecedoble](https://user-images.githubusercontent.com/5034215/44725173-64c61e80-aadd-11e8-849e-4b9c868d3d57.png)
In other quotations it appears correctly:
![normalporaqui](https://user-images.githubusercontent.com/5034215/44725198-70194a00-aadd-11e8-932c-40fd4c2827ec.png)
I cannot understand why happens this.

 Second one, we are thinking about show Sale Orders also in Analytic Account's view, as `mrp_analytic` do, but i think it needs a separetely project, what do you think about this?

Finally, we need someone that update the localization files, i'm noob in Odoo and i don't know how can i do that, and someone that update the test.

Thank you very much!!!

SalahAdDin committed:
:sunny: Added button for Analytics for Sale Orders view.

SalahAdDin committed:
Merge pull request #2 from SalahAdDin/11.0

:sunny: Added button for Analytics for Sale Orders view.

SalahAdDin commented:
TODO:

- [x] Show quotation(s) in the Project description.
- [ ] Refactor tests. 
- [ ] Review the Project Overview's button problem in Sale Order's view: it is missing.
- [x] Add a Sale Order's button in Account Analytic view (manufacture project module use it, it could be useful).

SalahAdDin commented:
Ok, guys, this is our work in order to update this module.

But we have two or three problems:
First one, the `Project Overview` button is showing in wrong places and sometimes in the same place with  `Create Project` button:
![aquiaparecedoble](https://user-images.githubusercontent.com/5034215/44725173-64c61e80-aadd-11e8-849e-4b9c868d3d57.png)
In other quotations it appears correctly:
![normalporaqui](https://user-images.githubusercontent.com/5034215/44725198-70194a00-aadd-11e8-932c-40fd4c2827ec.png)
I cannot understand why happens this.

 The second one, we are thinking about show Sale Orders also in Analytic Account's view, as `mrp_analytic` do, but i think it needs a separate project, what do you think about this?

Finally, we need someone that update the localization files, I'm a noob in Odoo and I don't know how can i do that, and someone that update the test.

Thank you very much!!!

SalahAdDin commented:
**Update:**
I found the problem with the `Project(s) Overview` button: https://github.com/odoo/odoo/blob/11.0/addons/sale_timesheet/views/sale_order_views.xml#L19

`sale_timesheet` module does not offer all the features that this module offers, so, in which why must we integrate both of them?

SalahAdDin committed:
:fire: fixed a problem with compute field's function.

SalahAdDin committed:
Merge pull request #3 from SalahAdDin/11.0

:fire: fixed a problem with compute field's function.

SalahAdDin commented:
:up: :+1:

nikul-serpentcs commented on sale_order_project/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+==================
+Link module to map sale orders to project
+        """,
+    'website': 'http://www.akretion.com/',
```
@aitorbouzas Replace URL as per [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L10)

nikul-serpentcs commented on sale_order_project/models/sale.py:
```diff
@@ -0,0 +1,59 @@
+from datetime import date
+
+from odoo import api, fields, models, _
+from odoo.exceptions import Warning
```
Remove `Warning`, Use `UserError`

nikul-serpentcs commented on sale_order_project/views/analytic_account_view.xml:
```diff
@@ -0,0 +1,24 @@
+        <field name="arch" type="xml">
+            <xpath expr='//div[@name="button_box"]' position='inside'>
+                <button name="%(sale.action_orders)d"
+                        icon="fa-shopping-cart"
```
remove `icon` and use `class`

ex.` <i class="fa-shopping-cart" />`

SalahAdDin replied:
i'm reverting that:
![que](https://user-images.githubusercontent.com/5034215/46792300-bb18b500-cd4b-11e8-99ab-19d5191240be.png)

nikul-serpentcs commented on sale_order_project/views/project_view.xml:
```diff
@@ -0,0 +1,20 @@
+                <button class="oe_stat_button"
+                        type="object"
+                        name="sale_order_tree_view"
+                        icon="fa-shopping-cart"
```
same here

SalahAdDin replied:
Reverting...

nikul-serpentcs reviewed (commented):
Improve Code

SalahAdDin commented:
@moylop260 hi, one review here please.
@nikul-serpentcs i did,

aitorbouzas commented:
@SalahAdDin even if they don't review you can fix the lints shown in the travis bot.

SalahAdDin commented:
@aitorbouzas Yes, i saw it, i'm it.

nikul-serpentcs commented on sale_order_project/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+    'website': 'https://github.com/OCA/project',
+    'depends': [
+        'project',
+        'sale',
```
@aitorbouzas IMO remove `sale`, because `sale_management` already depends on `sale`, so need here.

nikul-serpentcs commented on sale_order_project/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+    'category': 'Project',
+    'summary': 'Create a Project from Sale Order',
+    'license': 'AGPL-3',
+    'author': 'Akretion, AvanzOSC, Serv. Tecnol. Avanzados - Pedro M. Baeza, Adaptive City - Aitor Bouzas, José Luis Sandoval Alaguna, Odoo Community Association (OCA)',
```
@aitorbouzas too log line

nikul-serpentcs commented on sale_order_project/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+        'sale',
+        'sale_management',
+    ],
+    'data': [
```
@aitorbouzas Manage sequence should be [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L29)

SalahAdDin replied:
There is a problem with this sequence, the view is loaded before to the wizard, then, the view cannot read the registered action and that's why it does not work properly, i think.

In local it is not working.

nikul-serpentcs commented on sale_order_project/models/analytic_account.py:
```diff
@@ -0,0 +1,16 @@
+class AnalyticAccount(models.Model):
+    _inherit = "account.analytic.account"
+
+    def _compute_num_sale_orders(self):
```
I think here use `@api.depends` also.

SalahAdDin replied:
depends on what?

nikul-serpentcs replied:
Yes, I know nothing depends here then use only `@api.multi`.

SalahAdDin replied:
Ok, Multi, one moment.

SalahAdDin replied:
Done.

nikul-serpentcs commented on sale_order_project/models/project.py:
```diff
@@ -0,0 +1,34 @@
+class Project(models.Model):
+    _inherit = 'project.project'
+
+    def _compute_sale_orders_count(self):
```
same here

SalahAdDin replied:
Done

nikul-serpentcs commented on sale_order_project/views/project_view.xml:
```diff
@@ -0,0 +1,20 @@
+        <field name="model">project.project</field>
+        <field name="inherit_id" ref="project.edit_project"/>
+        <field name="arch" type="xml">
+            <xpath expr="//button[@name='attachment_tree_view']" position="after">
```
here directly use `button`

Ex. `<button name="attachment_tree_view" position="after">`

SalahAdDin replied:
`Element '<button name="attachment_tree_view">' cannot be located in parent view`.

nikul-serpentcs commented on sale_order_project/views/sale_view.xml:
```diff
@@ -0,0 +1,33 @@
+                    <field name="related_project_id"/>
+                </group>
+            </group>
+            <xpath expr="//button[@name='action_view_invoice']" position="before">
```
same here

nikul-serpentcs commented on sale_order_project/views/sale_view.xml:
```diff
@@ -0,0 +1,33 @@
+                </group>
+            </group>
+            <xpath expr="//button[@name='action_view_invoice']" position="before">
+                <!-- icon="fa-project-diagram"
```
Is need?

nikul-serpentcs commented on sale_order_project/wizard/sale_order_createproject.py:
```diff
@@ -0,0 +1,42 @@
+        if not project.id and not order.related_project_id:
+            # create new project.project
+            order.action_create_project()
+        # if project but order is not related to that project
```
IMO Remove dead code everywhere

SalahAdDin replied:
Which dead code? they are comments, can you explain me it, please? :D

aitorbouzas replied:
@SalahAdDin commented code should be out of the modules. I think that is what he means.

SalahAdDin replied:
Oh, ok, i'm deleting it right now.

nikul-serpentcs commented on sale_order_project/views/project_view.xml:
```diff
@@ -0,0 +1,20 @@
+        <field name="inherit_id" ref="project.edit_project"/>
+        <field name="arch" type="xml">
+            <xpath expr="//button[@name='attachment_tree_view']" position="after">
+                <!--icon="file-invoice-dollar"-->
```
Is need?

nikul-serpentcs reviewed (commented):
Improve Code

aitorbouzas commented:
@anajuaristi @oihane any news on your refactoring proposal?

@SalahAdDin is working on this, if we are going to dismiss this PR when it gets green on travis an runbot we better know now before working more on it.

SalahAdDin commented:
I don't understand, why has it this problem?
```
    raise ValueError('External ID not found in the system: %s' % xmlid)
odoo.tools.convert.ParseError: "External ID not found in the system: sale_order_project.sale_order_createproject_action" while parsing /home/odoo/build/OCA/project/sale_order_project/views/sale_view.xml:3, near
<record id="view_order_form" model="ir.ui.view">
        <field name="name">sale.order.project.form</field>
        <field name="model">sale.order</field>
        <field name="inherit_id" ref="sale.view_order_form"/>
        <field name="arch" type="xml">
            <data><group name="technical" position="after">
                <group name="project" string="Project">
                    <field name="related_project_id"/>
                </group>
            </group>
            <button name="action_view_invoice" position="after">
                <button class="oe_stat_button" type="action" name="%(sale_order_createproject_action)d" icon="fa-tasks" help="Create Project" groups="sale.group_analytic_accounting,project.group_project_user" attrs="{'invisible':['|',('state','!=','draft'),('related_project_id','!=',False)]}">
                    <div class="o_field_widget o_stat_info">
                        <span class="o_stat_text">Create</span>
                        <span class="o_stat_text">Project</span>
                    </div>
                </button>
            </button>
        </data></field>
    </record>
Message not found: 'Modules loaded.'
```

aitorbouzas commented:
@SalahAdDin your action called ```sale_order_createproject_action``` is being referenced in your ```sale_view.xml``` before being loaded by the ```sale_order_createproject_view.xml``` file. Change the order in the ```__manifest__.py```

SalahAdDin commented:
@aitorbouzas i knew it: https://github.com/OCA/project/pull/395#discussion_r224333819

SalahAdDin commented:
Ok, **runbot** passed, but, what's the problem with **travis**?

tafaRU commented:
> what's the problem with **travis**?

have a look at https://travis-ci.org/OCA/project/jobs/440026426#L588-L596

SalahAdDin commented:
@tafaRU Thank you man, i didn't see them. Done.

SalahAdDin committed:
[MIG][11.0][sale_order_project] Migration to 11.0

SalahAdDin commented:
Guys, it passed, but, i don't know how cover the other two test, is the two first test enough to merge this pull request?

aitorbouzas commented:
@SalahAdDin I would try to cover everything. Anyway, I don't know if this module is going to be merged after everything is green.

aitorbouzas commented on sale_order_project/models/analytic_account.py:
```diff
@@ -0,0 +1,20 @@
+# © 2018 José Luis Sandoval Alaguna - Rotafilo
```
Replace copyright symbol for ```Copyright```.

aitorbouzas commented on sale_order_project/models/sale.py:
```diff
@@ -0,0 +1,62 @@
+# © 2014 Akretion - Sébastien BEAU <sebastien.beau@akretion.com>
```
Same here

aitorbouzas commented on sale_order_project/models/sale.py:
```diff
@@ -0,0 +1,62 @@
+# © 2014 Akretion - Sébastien BEAU <sebastien.beau@akretion.com>
+# © 2014 Akretion - Benoît GUILLOT <benoit.guillot@akretion.com>
```
And here

aitorbouzas reviewed (commented):
A little comment about copyright symbol.

aitorbouzas committed:
Add tests

aitorbouzas commented:
@SalahAdDin added some tests, but I don't have more time. Check if you can do the rest.

aitorbouzas committed:
Fix flake

moylop260 commented:
After CI green could you squash the commits by author after migration ones, please?

SalahAdDin commented:
@aitorbouzas Can you do it? i'm not sure about it, i mean, i don't know how.

SalahAdDin committed:
:sunny: Add missing test for models.

SalahAdDin commented:
@aitorbouzas I'm sucks with test, we need to solve the final model test and the wizard function.

SalahAdDin committed:
:fire: trying with test.

aitorbouzas commented:
@SalahAdDin I see you are doing some progress. 
Do you know how to run the tests locally? Use ```--test-enable``` when running your server, this way you can test them yourself locally.
Edit: User -> Use

aitorbouzas commented:
Also you could add some pylint like flake and pep in your environment. This way you can check everything's correct before committing and pushing.

SalahAdDin commented:
I'm using `PyCharm` but it has no `flake8`. I'm using `docker` to run my server.

But my last trying for making test are so bad.

aitorbouzas commented:
You can install flake8 as an external tool in PyCharm and then you can use it in a folder. It'll check all the lints.
To run the test I do these steps:
1. Delete database
2. Run server with --test-enable -d test -i ```module_to_install```
3. Watch logs

moylop260 commented on sale_order_project/tests/test_sale_order_project.py:
```diff
@@ -0,0 +1,61 @@
+        self.wizard.action_create_project()
+
+    def test_error_action_create_project_task(self):
+        with self.assertRaises(UserError)
```
https://travis-ci.org/OCA/project/jobs/441667575#L1725
```bash
  File "/home/travis/build/OCA/project/sale_order_project/tests/test_sale_order_project.py", line 59
    with self.assertRaises(UserError)
                                    ^
SyntaxError: invalid syntax
```

You need to add `:` at final of this sentence.
and a mute_logger will be needed too, here.

moylop260 commented:
https://travis-ci.org/OCA/project/jobs/441667573#L651
`sale_order_project/__init__.py:1: [W8201(incoherent-interpreter-exec-perm), ] You have a python file with execution permissions but you don't have a interpreter magic comment. If you really needs a execution permission then add a magic comment ( https://en.wikipedia.org/wiki/Shebang_(Unix) ). If you don't needs a execution permission then remove it with: chmod -x __init__.py`

You can fixed them executing:
 - `find . -name "*.py" -exec chmod -x {} \;`

Since that the python file of a module doesn't require have execution permissions.

SalahAdDin committed:
Appliying review's changes.

SalahAdDin commented:
@moylop260 Can you review this again? I cannot see any new test results. Yesterday **Github** was having problems, at least for me.

SalahAdDin committed:
:fire: Reformat project name.

Each project creates a new analytic account for that project. Analytic Account's name will be: $project.name + $partnert.name

With the current name we have $partner.name + $sale.order.name + $year as name, so, the analytic account will have a duplicate identifier.

With this change the project name will be just $sale.order.name + $year.

SalahAdDin commented:
@moylop260 I'm sorry, i have no time now for looking at tests.
About the squashing, that i can do is "delete" all branch and remake it, but it will delete all @aitorbouzas commits.

rafaelbn commented:
Hello! Which is the difference between this module and Odoo Core functionality?

SalahAdDin commented:
Afaik, there is not a vanilla way to create projects from Sale Orders.

rafaelbn commented:
Well :thinking: here you are:

https://youtu.be/8C6_o0lCs-w

SalahAdDin commented:
@rafaelbn I cannot find the same behavior that you show in that video, what's your setup for getting some like that?
![deepin-screen-recorder_select area_20181217163304](https://user-images.githubusercontent.com/5034215/50090442-96dcc700-0219-11e9-91da-0e3b854bd580.gif)

rafaelbn commented:
Hi @SalahAdDin , 

https://www.odoo.com/documentation/user/11.0/sales/invoicing/time_materials.html

Regards

SalahAdDin commented:
@rafaelbn  did you look the mail in the mailing list? well, i think the vanilla workflow works just with product services, also in that page you can find it:

> From a product page set as a service, you will find two options under the invoicing tab, select both Timesheets on tasks and Create a task in a new project.

This module enables to create a new project, or choose a project without regarding wheter the product is a service or not.

rafaelbn commented:
OK!

> Hello! Which is the difference between this module and Odoo Core functionality?

- This module enables to create a new project, or choose a project without regarding wheter the product is a service or not.

Clarified

angelmoya commented on sale_order_project/models/sale.py:
```diff
@@ -0,0 +1,62 @@
+                    )
+                ))
+            vals = self._prepare_project_vals(order)
+            project = project_obj.create(vals)
```
why do not you use the function _create_analytic_account?

https://github.com/odoo/odoo/blob/11.0/addons/sale/models/sale.py#L574

If you use core functionality and create a project after confirm sale order, the project has name structure diferent than if you create with this module, there is not the best way.

SalahAdDin replied:
can you make a PR for this change?

aitorbouzas commented:
What should we do with this monster?

SalahAdDin commented:
We need help XD

SalahAdDin commented:
@aitorbouzas what did you do at final?
